cpp/src/arrow/c/bridge.cc

```diff
@@ -2152,6 +2211,27 @@ class ArrayStreamBatchReader : public RecordBatchReader {
   }
 
  private:
+  template <typename BatchT>
```

Comment:
Stuffing everything in one class makes for complicated code. Why not have a base class with common helpers and then two distinct derived classes?